implement requireSourceFieldsUsedExcept

nimatrueway opened the pull request:
An implementation of `requireSourceFieldsUsedExcept(_.field1, _.field2..)` as described in https://github.com/scalalandio/chimney/issues/248

TODO/Future Work:
- [x] move usedSourceFields out from DerivationResult state and move it to `ProductToProduct` class
- [ ] implement a vararg version of `def ignoreUnusedField(selector, selectors*)` as Mateus suggested
- [ ] add `FailOnSpecifiedUnused` subclass for `ActionOnUnused`
  - [ ] implement a new setting to DSL `def requireUsedField(selector, selectors*)` as Mateus suggested
  - [ ] add `RequireUsedField` to `TransformerOverrides` to add to list of fields to be checked if `FailOnSpecifiedUnused` is enabled

MateuszKubuszok commented on chimney/src/main/scala-2/io/scalaland/chimney/internal/compiletime/ChimneyTypesPlatform.scala:
```diff
@@ -406,6 +417,17 @@ private[compiletime] trait ChimneyTypesPlatform extends ChimneyTypes { this: Chi
+          A.asCtor[runtime.PathList.List[?, ?]]
+            .map(A0 => A0.param_<[runtime.Path](0) -> A0.param_<[runtime.PathList](1))
+      }
+    }
```
A list of paths would not be needed if instead this would apply a config, as many times as many paths there is in vararg. This would make less code to maintain in DSL and macros.

MateuszKubuszok reviewed (changes requested):
Hello, thank you for your PR!

I think it's really nice first iteration, showing that you really looked around the codebase 😄 

In general, for the sake of backward compatibility and limiting the amount of new types, and flexibility I would like to suggest several things:

 * separation of the general policy (`IgnoreUnusedFields`/`FailOnUnusedFields`/`FailOnUnusedFieldStrict`/etc) from its exceptions (`IgnoreField[Path, Cfg]`, `FailOnUnusedField[Path, Cfg]`) both in DSL, type-level config and config within macros
 * that would allow changing the policy globally or in the whole scope
 * allowed/rejected fields could be appended one by one, matching current conventions for other customizations
   * if needed there could be 3 possible overrides: ignore field, reject unused field, remove override for field

So that would mean:

 * storing field policy as a `TransformerFlag` similar to `ImplicitTransformerPreference` (allowing us to add more policiies in the future)
   * e.g. `.unusedFieldPolicy(FailOnUnused)`
 * adding exceptions to list... maybe requiring non-empty list of overrides
   * e.g. `def ignoreUnusedField(selector, selectors*)`
 * things would be stored in a slightly different way internally, mostly using `runtimeOverrides: Vector[(Path, TransformerOverride)]` which is already managing propagation of configs based on paths

Such DSL would require a but more keystrokes but it would better fit Chimney's existing conventions.

nimatrueway commented:
> Hello, thank you for your PR!
> 
> I think it's really nice first iteration, showing that you really looked around the codebase 😄
> ...

My pleasure! Thank you so much for the quick review and good ideas. I initially had a similar thought in https://github.com/scalalandio/chimney/issues/601, it's nice to hear it wasn't too far off.

I'll start developing the API change 🤓🚧

nimatrueway commented:
Hey @MateuszKubuszok, I addressed the architectural change you proposed (and almost all your comments). Namely I split the API into a flag and a setting to override configuration per field.

```scala
// implementation:done -> tracks all fields and forces compile error if any left unused
.enableUnusedFieldPolicy(FailOnUnused)

// implementation:done -> untracks a single field
// TODO: how do I implement the vararg version without introducing PathList type? could use your direction here
.withIgnoreUnusedField(_.b)
```

To avoid a API change in the future and preserve backward compatibility, it's probably best to fix the API in this PR. So I tried to formulate your vision. Here's the domain terms and phrases I imagined makes sense for this feature. Please share if you have any suggestion.

- I would like to `track` certain fields and `fail` transformation derivation if any are `unused`.
- I would like to `track` certain fields and spit out `warning` during transformation derivation if any are `unused`.
- I would like to `track` all fields of the source type by default.
- I would like to `track` none of fields of the source type by default.
- I would like to `ignore` certain fields if `unused`.
- I would like to `track` certain fields if `unused`.

In that case I imagined the flag to be parameterized on two aspects fail/warn on unused and whether to track all fields or none of the fields by default.

```scala
// TODO: does this make sense? isn't it too verbose?
.enableUnusedFieldPolicy(FailOnUnused, TrackAll)
.enableUnusedFieldPolicy(WarnOnUnused, TrackNone)

// TODO: to manually track a bunch of fields
.withTrackUnusedFields(_.b, _.c)
```

MateuszKubuszok commented:
Hmm, the way I see it we have 2 possible policies for now:

 * ignore any unused field
 * fail the compilation on any field which was not explicitly marked as unused

and both would additionally fail the compilation if someone marked a field as unused but it was required.

In some more-or-less distant future, after #115, we might have additional policies:

 * ignore unused fields in fallback, failing on unused fields in the main value
 * fail the compilation for for any field, either in main value or fallback, what wasn't explicitly marked as unused
   * this could be problematic as I cannot think of a good DSL to mark such field as explicitly unused

Additionally, we have to keep in mind that:

 * even though we say `field`, actually right now Chimney makes no distinction between vals defined like:
   ```scala
   case class Foo(a: Int, b: String)
   ```

   and:
 
   ```scala
   case class Foo() {
     val a: Int = 0
     val b: String = "text"
   }
   ```

   and, some people would probably prefer to make the distinction which they want to ignore, but that distinction would have to be implemented somewhere in the future, so for now, we would have to keep the distinction in mind when naming things, e.g. by using name like "unused source values"

 * these flags are global, so whatever name we use, should not collide with whatever users' might use in their codebases, verbose names are less of an issue

 * we would have to add similar thing for target subtypes (when it's an enum/a subtype), so we have to make sure that the name indicates which is which

 * ATM Chimney does not support showing warnings - it's either `info` for logs, or `error` for reasons why macro cannot be expanded, we might add it in the future, as something like `policyReportingStrategy(PolicyCheckFailureIsWarning)`/`policyReportingStrategy(PolicyCheckFailureIsError)` but that would increase the scope of this PR too much.
   
---

Currently I think the API what would work the best would be something like:

```scala
// sealed abstract class UnusedFieldPolicy
// object FailOnIgnoredSourceVal extends UnusedFieldPolicy

src.into[Target]
.enableUnusedFieldPolicy(FailOnIgnoredSourceVal) // 1 value for now
.disableUnusedFieldPolicy // ignores unused fields, the default
.withFieldUnused(_.foo.bar.baz) // expect that field is unused:
                                // - fail compilation if it is needed
                                // - do NOT treat field as ignored when UnusedFieldPolicy is enabled
                                // - be na noop when UnusedFieldPolicy is disabled and field is not used
.transform
```

Then we could test it similarly to other flags:

 * in both `TotalTransformerProductSpec` and `PartialTransformerProductSpec`:
 * `group("flag .enableUnusedFieldPolicy")`
   * `test("should be disabled by default")`
   * `test("should inform user which unused fields failed")` 
   * `test("should ignore explicitly unused fields")`
   * `test("should not allow marking required fields as unused")`
 * `group("flag .disableUnusedFieldPolicy")`
  * `test("should disable globally enabled .enableUnusedFieldPolicy")`
 * tests would have to check that nesting case classes would also work

---

Adding:

 * distinction between warnings and errors
 * distinction between case class fields and other `val`s
 * support for UnusedTargetSubtypePolicy

would be a job for another PR, and not necessarily something to do soon. I am starting to think that even this change would require some refactoring to how `TransformationContext` works with regard to `Path`s in source, so I might need to do some refactoring, but I am short on time these days and I could do it :/

nimatrueway commented:
> Hmm, the way I see it we have 2 possible policies for now:
>
> ...

Thanks for sharing your insights Mateus, I skimmed through your comment twice and we are aligned. I need to carve up sometime on the weekend to hack it. 🚀

MateuszKubuszok commented:
Hey, once again thank you for your effort, I am sorry if you invested some more time to this, since I needed this functionality for another WIP PR (#614) and I had to started implementing it there. I am not sure how long it will take to finish that other PR, but it would be hard now to switch to another implementation.

nimatrueway commented:
> Hey, once again thank you for your effort, I am sorry if you invested some more time to this, since I needed this functionality for another WIP PR (#614) and I had to started implementing it there. I am not sure how long it will take to finish that other PR, but it would be hard now to switch to another implementation.

Sorry man, I switched job around the same time as I was working on this PR. No, I'm working in Ruby/Sorbet shop and slogging to catch up to this new world. _(Having to deal with Ruby, I can't put into words how much I missed Scala)_

Best of luck to you!
